// @stryker-mutator/jest-runner/src/configLoaders/ReactScriptsJestConfigLoader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const path_1 = tslib_1.__importDefault(require("path"));
const plugin_1 = require("@stryker-mutator/api/plugin");
const createReactJestConfig_1 = require("../utils/createReactJestConfig");
const pluginTokens_1 = require("../pluginTokens");
class ReactScriptsJestConfigLoader {
    constructor(resolve, projectRoot) {
        this.resolve = resolve;
        this.projectRoot = projectRoot;
    }
    loadConfig() {
        try {
            // Get the location of react script, this is later used to generate the Jest configuration used for React projects.
            const reactScriptsLocation = path_1.default.join(this.resolve('react-scripts/package.json'), '..');
            // Create the React configuration for Jest
            const jestConfiguration = this.createJestConfig(reactScriptsLocation);
            return jestConfiguration;
        }
        catch (e) {
            if (this.isNodeErrnoException(e) && e.code === 'MODULE_NOT_FOUND') {
                throw Error('Unable to locate package react-scripts. This package is required when projectType is set to "create-react-app".');
            }
            throw e;
        }
    }
    isNodeErrnoException(arg) {
        return arg.code !== undefined;
    }
    createJestConfig(reactScriptsLocation) {
        return createReactJestConfig_1.createReactJestConfig((relativePath) => path_1.default.join(reactScriptsLocation, relativePath), this.projectRoot, false);
    }
}
exports.default = ReactScriptsJestConfigLoader;
ReactScriptsJestConfigLoader.inject = plugin_1.tokens(pluginTokens_1.resolveToken, pluginTokens_1.projectRootToken);
