// @stryker-mutator/core/src/di/coreTokens.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.optionsValidator = exports.validationSchema = exports.pluginCreatorTestFramework = exports.pluginCreatorMutator = exports.pluginCreatorTestRunner = exports.pluginCreatorTranspiler = exports.pluginCreatorOptionsEditor = exports.pluginCreatorConfigEditor = exports.pluginCreatorReporter = exports.pluginDescriptors = exports.pluginKind = exports.reporter = exports.sandboxIndex = exports.transpiler = exports.loggingContext = exports.timeOverheadMS = exports.timer = exports.testFramework = exports.temporaryDirectory = exports.sandboxPool = exports.mutantTranspileScheduler = exports.transpiledFiles = exports.initialRunResult = exports.inputFiles = exports.configOptionsApplier = exports.configReader = exports.cliOptions = void 0;
exports.cliOptions = 'cliOptions';
exports.configReader = 'configReader';
exports.configOptionsApplier = 'configOptionsApplier';
exports.inputFiles = 'inputFiles';
exports.initialRunResult = 'initialRunResult';
exports.transpiledFiles = 'transpiledFiles';
exports.mutantTranspileScheduler = 'mutantTranspileScheduler';
exports.sandboxPool = 'sandboxPool';
exports.temporaryDirectory = 'temporaryDirectory';
exports.testFramework = 'testFramework';
exports.timer = 'timer';
exports.timeOverheadMS = 'timeOverheadMS';
exports.loggingContext = 'loggingContext';
exports.transpiler = 'transpiler';
exports.sandboxIndex = 'sandboxIndex';
exports.reporter = 'reporter';
exports.pluginKind = 'pluginKind';
exports.pluginDescriptors = 'pluginDescriptors';
exports.pluginCreatorReporter = 'pluginCreatorReporter';
exports.pluginCreatorConfigEditor = 'pluginCreatorConfigEditor';
exports.pluginCreatorOptionsEditor = 'pluginCreatorOptionsEditor';
exports.pluginCreatorTranspiler = 'pluginCreatorTranspiler';
exports.pluginCreatorTestRunner = 'pluginCreatorTestRunner';
exports.pluginCreatorMutator = 'pluginCreatorMutator';
exports.pluginCreatorTestFramework = 'pluginCreatorTestFramework';
exports.validationSchema = 'validationSchema';
exports.optionsValidator = 'optionsValidator';
