// @stryker-mutator/core/src/di/factoryMethods.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mutatorDescriptorFactory = exports.applyOptionsEditors = exports.loggerFactory = exports.testFrameworkFactory = exports.pluginResolverFactory = void 0;
const plugin_1 = require("@stryker-mutator/api/plugin");
const util_1 = require("@stryker-mutator/util");
const TestFrameworkOrchestrator_1 = require("../TestFrameworkOrchestrator");
const _1 = require(".");
function pluginResolverFactory(injector) {
    const pluginLoader = injector.injectClass(_1.PluginLoader);
    pluginLoader.load();
    return pluginLoader;
}
exports.pluginResolverFactory = pluginResolverFactory;
pluginResolverFactory.inject = plugin_1.tokens(plugin_1.commonTokens.injector);
function testFrameworkFactory(injector) {
    return injector.injectClass(TestFrameworkOrchestrator_1.default).determineTestFramework();
}
exports.testFrameworkFactory = testFrameworkFactory;
testFrameworkFactory.inject = plugin_1.tokens(plugin_1.commonTokens.injector);
function loggerFactory(getLogger, target) {
    return getLogger(target ? target.name : 'UNKNOWN');
}
exports.loggerFactory = loggerFactory;
loggerFactory.inject = plugin_1.tokens(plugin_1.commonTokens.getLogger, plugin_1.commonTokens.target);
function applyOptionsEditors(options, optionsEditorApplier) {
    optionsEditorApplier.edit(options);
    return util_1.deepFreeze(options);
}
exports.applyOptionsEditors = applyOptionsEditors;
applyOptionsEditors.inject = plugin_1.tokens(plugin_1.commonTokens.options, _1.coreTokens.configOptionsApplier);
function mutatorDescriptorFactory(options) {
    const defaults = {
        plugins: null,
        name: 'javascript',
        excludedMutations: [],
    };
    if (typeof options.mutator === 'string') {
        return Object.assign(Object.assign({}, defaults), { name: options.mutator });
    }
    return Object.assign(Object.assign({}, defaults), options.mutator);
}
exports.mutatorDescriptorFactory = mutatorDescriptorFactory;
mutatorDescriptorFactory.inject = plugin_1.tokens(plugin_1.commonTokens.options);
